Add Markdown linter

ilyatitovich opened the pull request:
Fixes #97

<!-- Describe what did you change -->

When searching for a Markdown linter, I settled on [remark](https://github.com/remarkjs/remark). The linter has an extensive collection of plugins and presets for processing Markdown files, allowing us to check only the necessary rules. It is possible to write custom rules to perform specific tasks.

I added 4 packages:

- [remark-cli](https://www.npmjs.com/package/remark-cli): CLI linter interface
- [remark-lint-no-dead-urls](https://www.npmjs.com/package/remark-lint-no-dead-urls): plugin that checks whether links are live
- [remark-validate-links](https://www.npmjs.com/package/remark-validate-links): plugin that checks links to files and folders
- [remark-lint-code-block-style](https://www.npmjs.com/package/remark-lint-code-block-style): plugin that checks code blocks and can format the file if necessary. To format files during linting, you need to add `--output` to the script: `"test:markdown": "remark . --output"`.

We may also need the [remark-toc](https://github.com/remarkjs/remark-toc) plugin, which allows us to generate TOC for *.md files.

Additionally, a `.remarkrc.js` file has been added, in which the necessary plugins are connected.

Using this linter, we can check whether links in a document are alive. For other tasks we need to write our own rules.

If this linter suits us, then I’ll start to work with GitHub Actions to add .github/workflows/docs.yml to run linter on **/*.md files changes.

And I'm a little confused by the task: `Find way to avoid calling in .github/workflows/root.yml`, because I don't see this file in the `workflows` directory.

## Motivation

Add to the project a good, flexible linter for checking *.md files for mistakes.

<!-- Why did you make these changes? -->

## Checklist

- [x] Don’t rush. Check all changes in PR again.
- [x] Think about changing documentation.
  - If you added a script to `scripts/`, add a comment with a description.
  - If you added a new folder, add its description to the project’s `README.md`.
  - If you added config, describe how we use this tool in the config’s comment.
  - If you added something to the project’s architecture, describe it in the project’s `README.md`.
  - Try to focus on “why?”, not “how?”.
- [x] If you added a new dependency, check our [requirements](../README.md#dependencies).
- [ ] Think about testing
  - If you added a feature, add unit tests.
  - If you added a new state to the UI, add visual tests.
  - If you fixed the bug, think about preventing bug regression in the future.
- If you changed web client:
  - [ ] Think about moving code to `core/`. What code will also be useful on other platforms?
  - [ ] Run `pnpm size` and check the difference in the JS bundle size. Is it relevant to the changes? Change the limit in `web/.size-limit.json` if necessary.
  - [ ] Think about keyboard UX. Is it easy to use the new feature with only one hand on a keyboard? Is it easy to understand what keys to press?
  - [ ] Think about HTML semantics.
  - [ ] Think about accessibility. Check a11y recommendations. Think about how screen reader users will use the tool. Is it easy to use on a screen with bad contrast?
  - [ ] Think about translations. Will
  - [ ] Think about right-to-left languages. What parts of the screen should be mirrored for Arabic or Hebrew languages?
- If you changed the colors token in the web client:
  - [ ] Think about app loading styles inlined in `index.html`.
- If you changed `core/`:
  - [ ] Think about making types more precise. Can you better explain data relations by type?
  - [ ] Think about conflict resolution. We don’t need some very smart changing merging; just 2 changes of the same item on different clients should not break the database. What if the user changes an item on one machine and removes it on another?
  - [ ] Think about log and storage migration.
- If you changed English translations:
  - [ ] Change translation ID if you change the meaning of the text.

ai commented on .remarkrc.js:
```diff
@@ -0,0 +1,12 @@
+  ]
+}
+
+export default preset
```
We can directly export object without saving it to variable

ai commented on .remarkrc.js:
```diff
@@ -0,0 +1,12 @@
+// File for setting up the markdown linter configuration.
```
The descriptions should be ready for junior developers who don’t remember `linter` and `Markdown` words.

Something like:

```
The config of tool to check documentation in .md files for popular mistakes
```

ai commented on package.json:
```diff
@@ -19,6 +19,7 @@
     "test:focused": "tsx ./scripts/check-focused-tests.ts",
     "test:messages": "tsx ./scripts/check-messages.ts",
     "test:versions": "tsx ./scripts/check-versions.ts",
+    "test:markdown": "remark .",
```
CI shows wrong warning

```
158:5-158:83 warning Cannot find file `(https:/github.blog/2019-02-14-introducing-draft-pull-requests/)` missing-file remark-validate-links:missing-file
```

Maybe it is a reason why PR is draft. Just note your.

ai replied:
Also try to make a mistake to docs and run `pnpm test:markdown`. Script should fail with error (not just print warning).

Try to break HTTP link (remove any symbol to make it dead), break local files link, etc.

ai commented on .remarkrc.js:
```diff
@@ -0,0 +1,12 @@
+  plugins: [
+    'remark-lint-no-dead-urls',
+    'remark-validate-links',
+    'remark-lint-code-block-style'
```
Does it just check between <code>```</code> and indent code block format?

We do not need it since most users don’t know about indent based code blocks.

I told you about a different thing. If you have a code block in *the list* you need to add indent to <code>```</code> block.

Wrong:

<pre>
1. List item
```
code
```
3. List item
</pre>

Right:

<pre>
1. List item
   ```
   code
   ```
3. List item
</pre>

ai commented:
> We may also need the [remark-toc](https://github.com/remarkjs/remark-toc) plugin, which allows us to generate TOC for *.md files.

We may need to write custom rule here too, since we need not generate TOC but to test current TOC (i am not sure how this plugin works to be honest, maybe it checks).

> And I'm a little confused by the task: Find way to avoid calling in .github/workflows/root.yml, because I don't see this file in the workflows directory.

Oops. I renamed this file to `test.yml` since I created an issue.

Here is a problem. We run `test.yml` on [non-md changes](https://github.com/hplush/slowreader/blob/main/.github/workflows/test.yml#L6-L9) because it runs a lot of expensive test (like staging deploy).

We need to keep separation between “just test docs quickly on spellchecking PR” and “change a code”. But `test.yml` now runs *all tests* which started from `test:`. So we need to change `test.yml` to not run `test:markdown` (but run it on `pnpm test` which should test *everything*).

ai commented on package.json:
```diff
@@ -40,6 +41,10 @@
     "postcss-html": "1.6.0",
     "prettier": "3.2.5",
     "prettier-plugin-svelte": "3.2.2",
+    "remark-cli": "^12.0.0",
```
Don’t forget to change Tools section in `README.md`

ilyatitovich commented:
In order to exclude the running of the `"test:markdown"` script from GitHub Actions, I suggest adding the `grep -v "test:markdown"` filter. It will skip and run all scripts starting with `"test:"` except `"test:markdown"`.

Added `docs.yml` file to run tests specifically for .md files.

I removed the `remark-lint-code-block-style` package, because it really doesn’t do what we need. I have not yet discovered a plugin that checks indentation inside a list element.

Added `--frail` to the script: `"test:markdown": "remark . --frail"`. Now linter displays a warning and returns an error:

![links](https://github.com/hplush/slowreader/assets/146111474/836d035b-3d4b-4214-a639-071360dd1cbf)


Added the remark tool to the **Tools** section in README.md.

ai commented on .github/workflows/docs.yml:
```diff
@@ -0,0 +1,18 @@
+name: Markdown Tests
```
```suggestion
name: Markdown Linter
```

ai commented on .github/workflows/docs.yml:
```diff
@@ -0,0 +1,18 @@
+on:
+  push:
+    branches:
+      - main
```
This workflow will not run on PR (and we need this linter to reduce review for PRs).

Copy this block from `test.yml` https://github.com/hplush/slowreader/blob/main/.github/workflows/test.yml#L10

ai commented on .github/workflows/docs.yml:
```diff
@@ -0,0 +1,18 @@
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check out code
+        uses: actions/checkout@v2
```
Very old action version

ai commented on .github/workflows/docs.yml:
```diff
@@ -0,0 +1,18 @@
+      - name: Check out code
+        uses: actions/checkout@v2
+      - name: Install Markdown linter and plugins
+        run: pnpm install remark-cli remark-lint-no-dead-urls remark-validate-links --ignore-scripts
```
Every time we will have different versions of `remark` here.

I suggest using pnpm lock file and just use the standard step with installing Node.js and all pnpm dependencies.

ai commented on .github/workflows/test.yml:
```diff
@@ -26,7 +26,7 @@ jobs:
         run: pnpm install --ignore-scripts
       - name: Run tests
-        run: pnpm test
+        run: pnpm test | grep -v "test:markdown"
```
`grep -v` will just hide `test:markdown` output. But we will still spend CPU on Markdown linter and CI will still fails on Markdown issues (but without a proper output why).

ai commented on .remarkrc.js:
```diff
@@ -0,0 +1,6 @@
+// Configuration file used to set up a tool to find mistakes in .md files.
```
```suggestion
// Configuration file used to find popular mistakes in .md files.
```

Let’s make it shorter.

ai commented on .remarkrc.js:
```diff
@@ -0,0 +1,6 @@
+// Configuration file used to set up a tool to find mistakes in .md files.
+// Add the name of the plugin you want to use to the list of plugins.
```
Other configs doesn’t have these description

```suggestion
```

ilyatitovich commented:
Updated the `docs.yml` file, adding PR bypass and standard installation of the environment and packages.

Updated the script in the `test.yml` file for running tests by adding the regular expression `'/^test:(?!markdown\b)/'`. It will allow us to run all tests except the "test:markdown". Checked locally - it works.

ai commented on .github/workflows/docs.yml:
```diff
@@ -0,0 +1,29 @@
+      - '**/*.md'
+  pull_request:
+    paths-ignore:
+      - './**/*.md'
```
This workflow will ignore all PRs with `*.md` changes. Seems like it is a mistake.

ai commented on .github/workflows/docs.yml:
```diff
@@ -0,0 +1,29 @@
+      - name: Run tests
+        run: pnpm test:markdown
+        env:
+          FORCE_COLOR: 1
```
We do not need `FORCE_COLOR` here. In `pnpm test` we need it because `pnpm` has an issue with nested tasks.

ai commented on .github/workflows/docs.yml:
```diff
@@ -0,0 +1,29 @@
+      - './.husky/**'
+jobs:
+  test:
+    permissions:
```
If we have a single `job`, let’s move `permission` to the root level as we have in another workflows with one job.

ai commented on README.md:
```diff
@@ -119,6 +119,7 @@ Global development tools:
 - [Prettier](./.prettierrc) to use the same code style formatting.
 - [TypeScript](./tsconfig.json) for strict type checking.
 - [ESLint](./eslint.config.js) to check for popular mistakes in JavaScript.
+- [remark](./.remarkrc.js) to find mistakes in .md files.
```
```suggestion
- [remark](./.remarkrc.js) to find mistakes in `.md` files.
```

1. Better formatting is better for readability.
2. File names should be in `<code>` according to practice in the web.
3. Without `<code>` the filename dot conflicts with typography dot.

ai commented:
1. When you will fix issues, press `Resolve conversation`
2. We need also to resolve conflicts

But I think it is the last code review. We are very close to merging.

ilyatitovich commented:
Corrected logical mistakes in `docs.yml`. Resolved conflicts.

Interesting task! I learned a lot of new things :)
